src/mesh_tools.rs

```diff
@@ -2,15 +2,165 @@ use std::collections::{HashMap, HashSet};
+            }
+
+            // Find the indices of edges of the current face in the list of edges,
+            // from which the topology was created
+            let face_unoriented_edges = proper_current_face.to_unoriented_edges();
+            let face_unoriented_edge_indices =
+                face_unoriented_edges.iter().map(|unoriented_edge| {
+                    unoriented_edge_index_map
+                        .get(unoriented_edge)
+                        .expect("The current edge is not found in the edge collection")
+                });
+
+            // Convert into oriented edges so that the neighboring faces can check
+            // for correct winding
+            let face_oriented_edges: Vec<_> =
+                face_unoriented_edges.iter().map(|u_e| u_e.0).collect();
+            // For each face edge index
+            for (i, face_unoriented_edge_index) in face_unoriented_edge_indices.enumerate() {
+                // get the actual oriented edge
+                let face_oriented_edge = face_oriented_edges[i];
+                // and try to find it in the edge-to-face topology.
+                if let Some(edge_in_faces) = edge_to_face_topology.get(&face_unoriented_edge_index)
+                {
+                    // If it exists, iterate the faces containing the edge
+                    for face_index in edge_in_faces {
+                        // and if it was not already added to the stack or even checked
+                        if !face_treatment_pattern[cast_usize(*face_index)] {
+                            // add it to the stack with the expected edge orientation
+                            edge_face_stack.push((face_oriented_edge.to_reverted(), *face_index));
+                            // and mark it treated.
+                            face_treatment_pattern[cast_usize(*face_index)] = true;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // Rebuild the mesh geometry with the new faces and original faces and normals.
+    Geometry::from_triangle_faces_with_vertices_and_normals(
+        synchronized_triangle_faces,
+        geometry.vertices().to_vec(),
+        geometry.normals().to_vec(),
+    )
+}
+
+/// Reverts vertex and normal winding of all faces in the mesh geometry and
+/// returns a reverted mesh geometry
+pub fn revert_mesh_faces(geometry: &Geometry) -> Geometry {
+    let reverted_faces = geometry.faces().iter().map(|face| match face {
+        Face::Triangle(t_f) => t_f.to_reverted(),
+    });
+    Geometry::from_triangle_faces_with_vertices_and_normals(
+        reverted_faces,
+        geometry.vertices().to_vec(),
+        geometry.normals().to_vec(),
+    )
+}
 
 /// Crawls the geometry to find continuous patches of geometry.
 /// Returns a vector of new separated geometries.
 pub fn separate_isolated_meshes(geometry: &Geometry) -> Vec<Geometry> {
```

Comment:
Wasn't this already done differently? The topology was outside the function.
Nevermind, I'll do it again.